client/client-lib/src/api.rs

```diff
@@ -149,33 +159,66 @@ impl WsFederationApi<WsClient> {
-                            .build(url)
-                            .await
-                            .expect("unable to connect to server"),
-                    )
+            members: members
+                .into_iter()
+                .map(|(peer_id, url)| FederationMember {
+                    peer_id,
+                    url,
+                    client: RwLock::new(None),
                 })
-                .collect()
-                .await,
+                .collect(),
             max_evil,
         }
     }
 }
 
-impl<C: ClientT> WsFederationApi<C> {
+impl WsFederationApi<WsClient> {
     pub async fn request<P: serde::Serialize, R: serde::de::DeserializeOwned>(
         &self,
         method: &str,
         param: P,
     ) -> Result<R> {
-        let params = [serde_json::to_value(param).expect("encoding error")];
-        let mut requests = futures::stream::iter(&self.clients).then(|(_id, client)| {
-            client.request::<R>(method, Some(jsonrpsee_types::ParamsSer::ArrayRef(&params)))
+        let params = Arc::new([serde_json::to_value(param).expect("encoding error")]);
+        let mut requests = futures::stream::iter(&self.members).then(|member| {
+            let params = params.clone();
+            Box::pin(async move {
+                let params = Some(jsonrpsee_types::ParamsSer::ArrayRef(&*params));
+                let rclient = member.client.read().await;
+                match &*rclient {
+                    Some(client) if client.is_connected() => {
+                        return client.request::<R>(method, params).await;
+                    }
+                    _ => {}
+                };
+
+                warn!(peer = %member.peer_id, "web socket disconnected, reconnecting");
+
+                drop(rclient);
+                let mut wclient = member.client.write().await;
+                match &*wclient {
+                    Some(client) if client.is_connected() => {
+                        // other task has connected it again
+                        let rclient = RwLockWriteGuard::downgrade(wclient);
+                        rclient.as_ref().unwrap().request::<R>(method, params).await
+                    }
+                    _ => {
+                        // write lock is acquired before creating a new client
+                        // so only one task will try to create a new client
+                        match WsClientBuilder::default().build(&member.url).await {
+                            Ok(client) => {
+                                *wclient = Some(client);
+                                // drop the write lock before making the request
+                                let rclient = RwLockWriteGuard::downgrade(wclient);
+                                rclient.as_ref().unwrap().request::<R>(method, params).await
+                            }
+                            Err(err) => {
+                                error!(%err, "unable to connect to server");
```

Comment:
We should know the server we are trying to connect to here to make debugging easier, maybe also open a span in the beginning?

Reply:
just moved this into a separate function